File: dart_sample/lib/app_bar_actions_padding_demo_page.dart
import 'package:flutter/material.dart';

import 'counter_widgets.dart';

class AppBarActionsPaddingDemoPage extends StatefulWidget {
  const AppBarActionsPaddingDemoPage({super.key});

  @override
  State<AppBarActionsPaddingDemoPage> createState() =>
      _AppBarActionsPaddingDemoPageState();
}

class _AppBarActionsPaddingDemoPageState
    extends State<AppBarActionsPaddingDemoPage> {
  EdgeInsets _themeActionsPadding = const EdgeInsets.fromLTRB(6, 2, 6, 2);
  bool _useWidgetOverride = false;
  EdgeInsets _widgetActionsPadding = const EdgeInsets.fromLTRB(10, 4, 10, 4);

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 10,
      children: <Widget>[
        const Text(
          'AppBar theme actionsPadding',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'Runtime probe for actions row padding resolution: widget override beats appBarTheme, appBarTheme beats default zero padding.',
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'theme 0',
              onTap: () => _setThemeActionsPadding(EdgeInsets.zero),
              width: 92,
              background: const Color(0xFFDCE3ED),
            ),
            _buildButton(
              label: 'theme 6/2',
              onTap: () => _setThemeActionsPadding(
                const EdgeInsets.fromLTRB(6, 2, 6, 2),
              ),
              width: 92,
              background: const Color(0xFFDCE3ED),
            ),
            _buildButton(
              label: 'theme 14/6',
              onTap: () => _setThemeActionsPadding(
                const EdgeInsets.fromLTRB(14, 6, 14, 6),
              ),
              width: 102,
              background: const Color(0xFFDCE3ED),
            ),
          ],
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'override off',
              onTap: () => _setUseWidgetOverride(false),
              width: 108,
              background: const Color(0xFFE9F5EC),
            ),
            _buildButton(
              label: 'override on',
              onTap: () => _setUseWidgetOverride(true),
              width: 108,
              background: const Color(0xFFE9F5EC),
            ),
          ],
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'widget 4/0',
              onTap: () => _setWidgetActionsPadding(
                const EdgeInsets.fromLTRB(4, 0, 4, 0),
              ),
              width: 92,
              background: const Color(0xFFF3E8D8),
            ),
            _buildButton(
              label: 'widget 10/4',
              onTap: () => _setWidgetActionsPadding(
                const EdgeInsets.fromLTRB(10, 4, 10, 4),
              ),
              width: 92,
              background: const Color(0xFFF3E8D8),
            ),
            _buildButton(
              label: 'widget 16/8',
              onTap: () => _setWidgetActionsPadding(
                const EdgeInsets.fromLTRB(16, 8, 16, 8),
              ),
              width: 102,
              background: const Color(0xFFF3E8D8),
            ),
          ],
        ),
        Text(
          'theme=${_formatInsets(_themeActionsPadding)}, override=${_useWidgetOverride ? 'on' : 'off'}, widget=${_formatInsets(_widgetActionsPadding)}, effective=${_formatInsets(_effectiveActionsPadding())}',
          style: const TextStyle(fontSize: 12, color: Colors.blueGrey),
        ),
        Container(
          width: 300,
          color: const Color(0xFFE7EDF6),
          padding: const EdgeInsets.all(10),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            spacing: 8,
            children: <Widget>[
              const Text(
                'Themed app bar preview',
                style: TextStyle(fontSize: 11, color: Colors.black54),
              ),
              _buildThemedPreview(),
              const Text(
                'Default app bar reference (actionsPadding = 0)',
                style: TextStyle(fontSize: 11, color: Colors.black54),
              ),
              _buildDefaultReferencePreview(),
            ],
          ),
        ),
      ],
    );
  }

  Widget _buildThemedPreview() {
    return Theme(
      data: ThemeData.light().copyWith(
        appBarTheme: AppBarTheme(
          backgroundColor: const Color(0xFF1E3A5F),
          foregroundColor: Colors.white,
          actionsPadding: _themeActionsPadding,
        ),
      ),
      child: AppBar(
        title: const Text('Theme actions padding'),
        actionsPadding: _useWidgetOverride ? _widgetActionsPadding : null,
        actions: <Widget>[_buildActionBadge('A1'), _buildActionBadge('A2')],
      ),
    );
  }

  Widget _buildDefaultReferencePreview() {
    return AppBar(
      title: const Text('Default actions padding'),
      backgroundColor: const Color(0xFF1E3A5F),
      foregroundColor: Colors.white,
      actions: <Widget>[_buildActionBadge('A1'), _buildActionBadge('A2')],
    );
  }

  static Widget _buildActionBadge(String label) {
    return Container(
      width: 28,
      height: 22,
      color: const Color(0xFFFFB703),
      child: Center(
        child: Text(
          label,
          style: const TextStyle(fontSize: 10, color: Colors.black),
        ),
      ),
    );
  }

  Widget _buildButton({
    required String label,
    required VoidCallback onTap,
    required double width,
    required Color background,
  }) {
    return SizedBox(
      width: width,
      child: CounterTapButton(
        label: label,
        onTap: onTap,
        background: background,
        foreground: Colors.black,
        fontSize: 12,
        padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
      ),
    );
  }

  void _setThemeActionsPadding(EdgeInsets value) {
    setState(() {
      _themeActionsPadding = value;
    });
  }

  void _setUseWidgetOverride(bool value) {
    setState(() {
      _useWidgetOverride = value;
    });
  }

  void _setWidgetActionsPadding(EdgeInsets value) {
    setState(() {
      _widgetActionsPadding = value;
    });
  }

  EdgeInsets _effectiveActionsPadding() {
    return _useWidgetOverride ? _widgetActionsPadding : _themeActionsPadding;
  }

  static String _formatInsets(EdgeInsets insets) {
    return '${insets.left.toStringAsFixed(0)}/${insets.top.toStringAsFixed(0)}/${insets.right.toStringAsFixed(0)}/${insets.bottom.toStringAsFixed(0)}';
  }
}

File: src/Sample/Flutter.Net/AppBarActionsPaddingDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Flutter.Material;
using Flutter.Rendering;
using Flutter.Widgets;

// Dart parity source (reference): dart_sample/lib/app_bar_actions_padding_demo_page.dart (exact sample parity)

namespace Flutter.Net;

public sealed class AppBarActionsPaddingDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new AppBarActionsPaddingDemoPageState();
    }
}

internal sealed class AppBarActionsPaddingDemoPageState : State
{
    private Thickness _themeActionsPadding = new(6, 2, 6, 2);
    private bool _useWidgetOverride;
    private Thickness _widgetActionsPadding = new(10, 4, 10, 4);

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 10,
            children:
            [
                new Text("AppBar theme actionsPadding", fontSize: 20, color: Colors.Black),
                new Text(
                    "Runtime probe for actions row padding resolution: widget override beats appBarTheme, appBarTheme beats default zero padding.",
                    fontSize: 14,
                    color: Colors.DimGray),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("theme 0", () => SetThemeActionsPadding(new Thickness()), width: 92, colorHex: "#FFDCE3ED"),
                        BuildButton("theme 6/2", () => SetThemeActionsPadding(new Thickness(6, 2, 6, 2)), width: 92, colorHex: "#FFDCE3ED"),
                        BuildButton("theme 14/6", () => SetThemeActionsPadding(new Thickness(14, 6, 14, 6)), width: 102, colorHex: "#FFDCE3ED"),
                    ]),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("override off", () => SetUseWidgetOverride(false), width: 108, colorHex: "#FFE9F5EC"),
                        BuildButton("override on", () => SetUseWidgetOverride(true), width: 108, colorHex: "#FFE9F5EC"),
                    ]),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("widget 4/0", () => SetWidgetActionsPadding(new Thickness(4, 0, 4, 0)), width: 92, colorHex: "#FFF3E8D8"),
                        BuildButton("widget 10/4", () => SetWidgetActionsPadding(new Thickness(10, 4, 10, 4)), width: 92, colorHex: "#FFF3E8D8"),
                        BuildButton("widget 16/8", () => SetWidgetActionsPadding(new Thickness(16, 8, 16, 8)), width: 102, colorHex: "#FFF3E8D8"),
                    ]),
                new Text(
                    $"theme={FormatInsets(_themeActionsPadding)}, override={(_useWidgetOverride ? "on" : "off")}, widget={FormatInsets(_widgetActionsPadding)}, effective={FormatInsets(EffectiveActionsPadding())}",
                    fontSize: 12,
                    color: Colors.DarkSlateGray),
                new Container(
                    width: 300,
                    color: Color.Parse("#FFE7EDF6"),
                    padding: new Thickness(10),
                    child: new Column(
                        crossAxisAlignment: CrossAxisAlignment.Stretch,
                        spacing: 8,
                        children:
                        [
                            new Text("Themed app bar preview", fontSize: 11, color: Colors.DimGray),
                            BuildThemedPreview(),
                            new Text("Default app bar reference (actionsPadding = 0)", fontSize: 11, color: Colors.DimGray),
                            BuildDefaultReferencePreview(),
                        ])),
            ]);
    }

    private Widget BuildThemedPreview()
    {
        return new Theme(
            data: ThemeData.Light with
            {
                AppBarTheme = new AppBarThemeData(
                    BackgroundColor: Color.Parse("#FF1E3A5F"),
                    ForegroundColor: Colors.White,
                    ActionsPadding: _themeActionsPadding),
            },
            child: new AppBar(
                titleText: "Theme actions padding",
                actionsPadding: _useWidgetOverride ? _widgetActionsPadding : null,
                actions:
                [
                    BuildActionBadge("A1"),
                    BuildActionBadge("A2"),
                ]));
    }

    private Widget BuildDefaultReferencePreview()
    {
        return new AppBar(
            titleText: "Default actions padding",
            backgroundColor: Color.Parse("#FF1E3A5F"),
            foregroundColor: Colors.White,
            actions:
            [
                BuildActionBadge("A1"),
                BuildActionBadge("A2"),
            ]);
    }

    private static Widget BuildActionBadge(string label)
    {
        return new Container(
            width: 28,
            height: 22,
            color: Color.Parse("#FFFFB703"),
            child: new Center(
                child: new Text(label, fontSize: 10, color: Colors.Black)));
    }

    private Widget BuildButton(string label, Action onTap, double width, string colorHex)
    {
        return new SizedBox(
            width: width,
            child: new CounterTapButton(
                label: label,
                onTap: onTap,
                background: Color.Parse(colorHex),
                foreground: Colors.Black,
                fontSize: 12,
                padding: new Thickness(10, 8)));
    }

    private void SetThemeActionsPadding(Thickness value)
    {
        SetState(() => _themeActionsPadding = value);
    }

    private void SetUseWidgetOverride(bool value)
    {
        SetState(() => _useWidgetOverride = value);
    }

    private void SetWidgetActionsPadding(Thickness value)
    {
        SetState(() => _widgetActionsPadding = value);
    }

    private Thickness EffectiveActionsPadding()
    {
        return _useWidgetOverride ? _widgetActionsPadding : _themeActionsPadding;
    }

    private static string FormatInsets(Thickness insets)
    {
        return $"{insets.Left:0}/{insets.Top:0}/{insets.Right:0}/{insets.Bottom:0}";
    }
}
